test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java

```diff
@@ -430,6 +445,215 @@ public HapiApiSpec balancesChangeOnTokenTransfer() {
+						cryptoCreate(FIRST_USER).key("signingKeyFirstUser"),
+						tokenCreate(A_TOKEN)
+								.tokenType(TokenType.NON_FUNGIBLE_UNIQUE)
+								.supplyType(TokenSupplyType.INFINITE)
+								.initialSupply(0)
+								.supplyKey("supplyKey")
+								.treasury(TOKEN_TREASURY),
+						tokenAssociate(FIRST_USER, A_TOKEN)
+				)
+				.when(
+						mintToken(A_TOKEN, List.of(ByteString.copyFromUtf8("memo")))
+				)
+				.then(
+						cryptoTransfer(
+								movingUnique(1, A_TOKEN).between(TOKEN_TREASURY, FIRST_USER)
+						)
+								.signedBy(DEFAULT_PAYER)
+								.hasKnownStatus(INVALID_SIGNATURE)
+				);
+	}
+
+	public HapiApiSpec uniqueTokenTxnWithReceiverNotSigned() {
+		return defaultHapiSpec("UniqueTokenTxnWithOwnerNotSigned")
+				.given(
+						newKeyNamed("supplyKey"),
+						newKeyNamed("signingKeyTreasury"),
+						newKeyNamed("signingKeyFirstUser"),
+						cryptoCreate(TOKEN_TREASURY).key("signingKeyTreasury"),
+						cryptoCreate(FIRST_USER).key("signingKeyFirstUser"),
+						tokenCreate(A_TOKEN)
+								.tokenType(TokenType.NON_FUNGIBLE_UNIQUE)
+								.supplyType(TokenSupplyType.INFINITE)
+								.initialSupply(0)
+								.supplyKey("supplyKey")
+								.treasury(TOKEN_TREASURY),
+						tokenAssociate(FIRST_USER, A_TOKEN)
+				)
+				.when(
+						mintToken(A_TOKEN, List.of(ByteString.copyFromUtf8("memo")))
+				)
+				.then(
+						cryptoTransfer(
+								movingUnique(1, A_TOKEN).between(TOKEN_TREASURY, FIRST_USER)
+						)
+								.signedBy("signingKeyTreasury", DEFAULT_PAYER)
+								.via("cryptoTransferTxn"),
+						getTxnRecord("cryptoTransferTxn").logged(),
+						getReceipt("cryptoTransferTxn").logged()
+				);
+	}
+
+	public HapiApiSpec uniqueTokenTxnsAreAtomic() {
+		return defaultHapiSpec("UniqueTokenTxnsAreAtomic")
+				.given(
+						newKeyNamed("supplyKey"),
+						newKeyNamed("signingKeyTreasury"),
+						newKeyNamed("signingKeyFirstUser"),
+						cryptoCreate(FIRST_USER).key("signingKeyFirstUser"),
+						cryptoCreate(SECOND_USER),
+						cryptoCreate(TOKEN_TREASURY).key("signingKeyTreasury"),
+						tokenCreate(A_TOKEN)
+								.tokenType(TokenType.NON_FUNGIBLE_UNIQUE)
+								.supplyType(TokenSupplyType.INFINITE)
+								.initialSupply(0)
+								.supplyKey("supplyKey")
+								.treasury(TOKEN_TREASURY),
+						tokenCreate(B_TOKEN)
+								.initialSupply(100)
+								.treasury(TOKEN_TREASURY),
+						mintToken(A_TOKEN, List.of(ByteString.copyFromUtf8("memo"))),
+						tokenAssociate(FIRST_USER, A_TOKEN),
+						tokenAssociate(FIRST_USER, B_TOKEN),
+						tokenAssociate(SECOND_USER, A_TOKEN)
+				)
+				.when(
+						mintToken(A_TOKEN, List.of(ByteString.copyFromUtf8("memo")))
+				)
+				.then(
+						cryptoTransfer(
+								moving(100, B_TOKEN).between(TOKEN_TREASURY, FIRST_USER)
+						),
+						cryptoTransfer(
+								movingUnique(1, A_TOKEN).between(FIRST_USER, SECOND_USER)
+						)
+								.hasKnownStatus(INSUFFICIENT_TOKEN_BALANCE)
```

Comment:
In order to check that transfers are atomic, we must transfer both fungible and non-fungible tokens inside 1 transaction.

There must be only 1 `cryptotransfer` with 2 `movings` for the 2 types of tokens. After that it is important to verify the effect after execution -> `fungible tokens` must not be transferred; `nonfungible tokens` must not be transferred

Reply:
It's correct in E2E test and framework